vector: add support for vector type

QuerthDP opened the pull request:
This pull request is an implementation of vector data type similar to one used by Apache Cassandra.

The patch contains:
* valid CQL syntax for VECTOR type
* adjustment of built-in type hierarchy
* implementation of vector_type_impl class
* support for serialization and deserialization of vectors
* unit tests 

Closes #2

piodul commented on cql3/cql3_type.cc:
```diff
@@ -303,6 +303,47 @@ class cql3_type::raw_tuple : public raw {
+        return false;
+    }
+
+    virtual void freeze() override {
+        if (_type->supports_freezing()) {
+            _type->freeze();
+        }
+        _frozen = true;
+    }
```
Looks like freezing a vector does not have any effect:

```
cqlsh> create table ks.t (a int, b frozen<vector<list<int>, 32>>, c vector<frozen<list<int>>, 16>, primary key (a));
cqlsh> describe ks.t;

CREATE TABLE ks.t (
    a int PRIMARY KEY,
    b vector<list<int>, 32>,
    c vector<frozen<list<int>>, 16>
) WITH additional_write_policy = '99p'
    AND allow_auto_snapshot = true
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND cdc = false
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'chunk_length_in_kb': '16', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND memtable = 'default'
    AND crc_check_chance = 1.0
    AND default_time_to_live = 0
    AND extensions = {}
    AND gc_grace_seconds = 864000
    AND incremental_backups = true
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair = 'BLOCKING'
    AND speculative_retry = '99p';
```

What's interesting, it doesn't seem to freeze the inner type. I am not sure whether this was intended in the original implementation, but let's be compatible.

QuerthDP replied:
To match Apache Cassandra's implementation, freezing of inner elements has been removed.

piodul commented on types/types.cc:
```diff
@@ -1609,6 +1631,124 @@ static void validate_aux(const tuple_type_impl& t, View v) {
+    // vector name into
+    // "org.apache.cassandra.db.marshal.FrozenType(...)".
+    // Even when the vector is frozen.
+    // For more details see #4087
```
Should we refer to this issue? It was created long before the vector datatype was implemented in Cassandra.

QuerthDP replied:
After adjusting the implementation to match Apache Cassandra, reference to this issue has been removed as it became irrelevant.

piodul commented on types/types.cc:
```diff
@@ -1609,6 +1631,124 @@ static void validate_aux(const tuple_type_impl& t, View v) {
+
+    size_t len = 0;
+
+    if(type->value_length_if_fixed()) {
```
Nit: space after `if`

piodul commented on types/types.cc:
```diff
@@ -1609,6 +1631,124 @@ static void validate_aux(const tuple_type_impl& t, View v) {
+
+    auto old_elements = m.get_elements_type();
+    if (auto new_elements = old_elements->update_user_type(updated)) {
+          return std::make_optional(static_pointer_cast<const abstract_type>(
+        vector_type_impl::get_instance(*new_elements, m.get_dimension())));
```
Nit: indentation became really weird here

piodul reviewed (commented):
I know that this is work in progress, but leaving some comments anyway.

Please use `Fixes` instead of `Closes` in the PR description. We use "Fixes" for referring to issues that will be fixed by the PR; we use "Closes" in merge commits to close PRs when they are merged to master.

QuerthDP commented on cql3/cql3_type.cc:
```diff
@@ -303,6 +303,47 @@ class cql3_type::raw_tuple : public raw {
+    }
+
+    virtual bool supports_freezing() const override {
+        return true;
```
According to Apache Cassandra's RawVector implementation, although applying `frozen<...>` on a vector does not result in any effect, support for freezing should be allowed. 

[Apache Cassandra's RawVector](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/cql3/CQL3Type.java#L1033)

Jadw1 reviewed (commented):
The PR is still a draft, but you should squash "fix ..." commits with commits which introduce need to do the fix. 
Fix means there existed a bug in the code and now you are fixing it. You shouldn't do a bug/todo in one commit and fix in later in the same PR.

QuerthDP commented:
Changelog:

- squashed "fix..." commits into core commits implementing fixed features
- adjusted commit messages to be more accurate
- swapped vector_type_impl and CQL3 syntax implementation as the latter should use vector_type_impl
- squashed additional tests into their respective test commits

Jadw1 reviewed (approved):
The PR looks good. Please comment with a report from test.py run (run all tests).

janpiotrlakomy commented:
test report from running `./test.py --no-gather-metrics `:
```

Found 3333 tests.
Setting minio proxy random seed to 1734539163
Starting S3 proxy server on ('127.133.60.2', 9002)
================================================================================
[N/TOTAL]   SUITE    MODE   RESULT   TEST
------------------------------------------------------------------------------
[3333/3333] topology_experimental_raft   dev   [ PASS ] topology_experimental_raft.test_topology_recovery_basic.1
Stopping S3 proxy server
------------------------------------------------------------------------------
CPU utilization: 9.9%

```

piodul commented on cql3/Cql.g:
```diff
@@ -1621,7 +1621,7 @@ collectionLiteral returns [uexpression value]
     : '['
           ( t1=term { l.push_back(std::move(t1)); } ( ',' tn=term { l.push_back(std::move(tn)); } )* )?
-      ']' { $value = collection_constructor{collection_constructor::style_type::list, std::move(l)}; }
+      ']' { $value = collection_constructor{collection_constructor::style_type::list_or_vector, std::move(l)}; }
```
> Motivation of using collection_constructor for representing vectors before preparation is that the vectorLiterals are indistinguishable from listLiterals before submiting any context of column type.

I think that the commit message could benefit from restructuring. You should start with motivation first, then explain what you are doing in the commit. It is much easier to read about the changes that you introduce when you have all the necessary context to understand _why_ you do them, rather than taking them at face value at first and later trying to fill the gaps in your understanding by reading the motivation. Apart from that, I think that this order generally leads to more natural and easier to understand text.

Considering that I recommend to split this into two commits - add `vector_constructor` + rename `list` to `list_or_vector` (I think this order makes most sense) - I think that the descriptions should follow this structure:

Add `vector_constructor`:
- Explain why it makes sense to have a separate `vector_costructor` from `collection_constructor`
- Note that the expression data structures can represent both typed and untyped expressions
- Explain that we cannot use the new variant for untyped expressions because the literal syntax for lists and vectors is exactly the same
- Explain that the variant is effectively unused as of this commit, but the next commit will adjust `list` to `list_or_vector` and will make use of the new `vector_constructor`

Rename `list` to `list_or_vector`:
- No need to provide lengthy motivation, it is sufficient to refer to the previous commit ("as explained in the previous commit, ..." etc.)
- Shortly sum up what is being done in this commit

QuerthDP replied:
Proposed changes has been introduced during rebase.

Please check if the changes meet your requests.

piodul commented on test/boost/expr_test.cc:
```diff
@@ -7,6 +7,7 @@
 #include "types/types.hh"
 #include "types/list.hh"
 #include "types/map.hh"
+#include "types/vector.hh"
```
> boost/expr_test: add vector expression tests
> 
> I've added/adjusted tests using vector_constructor and collection_constructor with list_or_vector style before preparation.

Use imperative mood, please

> 
> Implemented utilities used in expr_test similar to those added in https://github.com/zpp-2024-vector-search/scylladb/commit/8f6309bd660084f802cae7746a5f9ed3bcaa4685

QuerthDP replied:
The commit message has been changed accordingly.

piodul commented on test/cqlpy/cassandra_tests/validation/entities/vectors_test.py:
```diff
@@ -0,0 +1,16 @@
+# This file was translated from the original Java test from the Apache
+# Cassandra source repository, as of commit 54e46880690bd5effb31116986292c1bdc9e891e
+#
+# The original Apache Cassandra license:
```
> cassandra_tests: translate tests covering the vector type.
> 
> Added cql_vector_test which tests the basic functionalities of the vector type using CQL.
> Added vectors_test which tests if descending ordering of vector is supported.

Use imperative mood, please

QuerthDP replied:
The commit message has been changed accordingly.

piodul reviewed (changes requested):
The code and the tests in the PR are comprehensive and should be generally fine. What should be improved is 1) commit messages 2) structure of the PR 3) documentation.

The standard formatting rule for the commit messages is: try to keep the title up to max 50 characters, then keep the lines up to 72 characters max. Please format your commit messages according to it (stick at least to the second rule).

Please use the imperative mood, or at least past tense in your commit messages. Try to avoid writing in first person when you say what your commit is doing - think about commits as entities which introduce changes themselves. So, "Add this and this to the code" is recommended. Some people (me included, sometimes) use the form "This commit adds this and this to the code", which is not imperative mood but better fits the narrative about commits doing things. It is fine to use first person when talking about the motivation behind the commit, though.

About the documentation: please look around in the documentation directory, but right away I can tell you that you will have to update at least `docs/cql/types.rst`. Please follow the instructions in `docs/README.md` to learn how to build the documentation locally (protip: if you encounter some keyring-related problems with running the poetry command, I found out that defininng the `PYTHON_KEYRING_BACKEND=keyring.backends.null.Keyring` environment variable helps).

QuerthDP commented on types/types.cc:
```diff
@@ -2851,6 +3044,25 @@ struct from_string_visitor {
         }
         return concat_fields(fields, field_len);
     }
+    bytes operator()(const vector_type_impl& t) {
+        std::vector<std::string_view> field_strings = split_field_strings(s, ',');
+        if (field_strings.size() > t.get_dimension()) {
+            throw marshal_exception(
+                    format("Invalid vector literal: too many elements. Type {} expects {:d} but got {:d}",
+                            t.as_cql3_type(), t.get_dimension(), field_strings.size()));
+        }
+        std::vector<bytes> fields(field_strings.size());
+        std::vector<int32_t> field_len(field_strings.size());
+        for (std::size_t i = 0; i < field_strings.size(); ++i) {
+            field_strings[i] = trim(field_strings[i]);
+            fields[i] = t.get_elements_type()->from_string(field_strings[i]);
+            field_len[i] = fields[i].size();
+        }
+        if (t.get_elements_type()->is_native()) {
+            return concat_fields(fields);
+        }
+        return concat_fields(fields, field_len);
+    }
```
It is tested [here](https://github.com/zpp-2024-vector-search/scylladb/blob/83604d7f9867771776c805fe82377439552141e3/test/boost/types_test.cc#L612C5-L617C56) similar to the tuple type.

piodul replied:
Can we change this test to be a cqlpy test, so that you can run it against Cassandra as well?

QuerthDP replied:
The test has been added via df2a84d4570978f74e4f5464ea5187e4f1590aa1.

Although Cassandra seems not to support such way of insertion, it works fine with our implementation.
The syntax for this was based on [to_string ](https://github.com/zpp-2024-vector-search/scylladb/blob/df2a84d4570978f74e4f5464ea5187e4f1590aa1/types/types.cc#L3094C1-L3156C3) function which acts the same for list, set and vector, and outputs their elements in a string split by commas.

Nevertheless I'm not sure now if this functionality should be added, if it is not supported the same way in Cassandra
(seems like it only works with tuples, I was not able to deduce the correct syntax for collections and vectors).
Furthermore collections don't implement it as well [(marked as FIXME)](https://github.com/zpp-2024-vector-search/scylladb/blob/df2a84d4570978f74e4f5464ea5187e4f1590aa1/types/types.cc#L3046C5-L3050C6).

Could you express your opinion on this case?

piodul replied:
> Although Cassandra seems not to support such way of insertion, it works fine with our implementation.

If Cassandra does not support this syntax then there is a risk that it will in the future, and will decide to do so in an incompatible way to Scylla's. In that case I'd prefer we didn't introduce support for it.

> [(marked as FIXME)](https://github.com/zpp-2024-vector-search/scylladb/blob/df2a84d4570978f74e4f5464ea5187e4f1590aa1/types/types.cc#L3046C5-L3050C6)

I don't think we should outright abort the node if we try it (by the way, can this abort be triggered now with collections?), but we should return a similar error to what Cassandra returns.

QuerthDP replied:
> I don't think we should outright abort the node if we try it (by the way, can this abort be triggered now with collections?), but we should return a similar error to what Cassandra returns.

As far as I've checked, trying to insert into a list or set column using string results in InvalidRequest exception **from the driver**, so the abort() function is not called as it is adressed by this exception before.

Although driver seems not to support vectors in such case, and allows the execution of from_string_visitor (leading to abort() function if we removed the implementation).

Furthermore, in Cassandra it works OK with vectors, the same exception as for lists and sets.

I'm still not sure how should we deal with this problem, further ideas highly appreciated.

QuerthDP replied:
As decided during status meeting (with @piodul, @Jadw1, @janpiotrlakomy being present), the from_string implementation alongside unnecessary utility function changes and from_string cqlpy test have been removed.

It has been adjusted to match Scylla's behaviour of trying to insert collections from string, which leads to throwing an exception (Invalid STRING constant ...) caused by untyped constant not being able to be assigned to neither vectors nor collections.

Furthermore the from_string_visitor implementation for vectors has been marked FIXME similar to from_string_visitor implementation for collections, as we believe both of them will be resolved in the future.

Although we are aware that this implementation (or rather lack of it) is not fully compatible with Apache Cassandra,
we decided to keep it that way as the functionality is truly niche, and should not be introduced in advance, while collections does not support it yet.

QuerthDP commented on test/boost/expr_test.cc:
```diff
@@ -1804,8 +1867,6 @@ BOOST_AUTO_TEST_CASE(prepare_list_collection_constructor_no_receiver) {
             },
         .type = nullptr};
 
-    data_type list_type = list_type_impl::get_instance(long_type, true);
-
```
The list_type variable was not used later on in this test. Felt like this line is added by mistake due to copy-pasting of other tests.
But surely I think now this change should not be included in this commit, as we do not change this test any other way.

QuerthDP replied:
Reverted in rebase.

janpiotrlakomy commented on test/cqlpy/cassandra_tests/validation/operations/cql_vector_test.py:
```diff
@@ -0,0 +1,478 @@
+        # test()
+
+        # execute(cql, table, "INSERT INTO %s (pk) VALUES ([1, ?])", 2)
+        # test()
```
It was commented out accidentally, it works.

janpiotrlakomy commented on test/cqlpy/cassandra_tests/validation/operations/cql_vector_test.py:
```diff
@@ -0,0 +1,478 @@
+        # assertInvalidThrowMessage(cql, table, "String didn't validate.",
+        #                           InvalidRequest,
+        #                           "INSERT INTO %s (pk, value) VALUES (0, ?)",
+        #                           [BigInteger.valueOf(Long.MAX_VALUE).add(BigInteger.ONE), BigInteger.ONE])
```
This one was also left not translated and commented out accidentally, like the rest of this test it should work when python driver adds support for variable width types in vectors.

janpiotrlakomy commented on types/types.cc:
```diff
@@ -1609,6 +1631,143 @@ static void validate_aux(const tuple_type_impl& t, View v) {
+    return intern::get_instance(elements, dimension);
+}
+
+bytes_view read_vector_element(bytes_view& v, std::optional<size_t> value_length_if_fixed) {
```
In the end the duplicated function wasn't needed, because the same result could be achieved by just creating a basic_single_fragmented_view from a bytes_view and using the version for fragmented views. I've added a commit that fixes this.

QuerthDP commented:
Changelog:

- fixed nitpicks
- changed commit messages to be more precise and to use imperative mood
- split vector expression related commit into separate introduction of vector constructor and collection constructor adjustments
- removed unnecessary vector constructor tests
- added vector type documentation
- moved cql3 type mapping documentation into docs commit
- reintroduced commented out (by mistake) cqlpy tests
- refactored read_vector_element not to be a duplicated function
- added from_string cqlpy test

QuerthDP commented on cql3/Cql.g:
```diff
@@ -1936,6 +1936,7 @@ comparator_type [bool internal] returns [shared_ptr<cql3_type::raw> t]
     : n=native_or_internal_type[internal]     { $t = cql3_type::raw::from(n); }
     | c=collection_type[internal]   { $t = c; }
     | tt=tuple_type[internal]       { $t = tt; }
+    | vt=vector_type[internal]      { $t = vt; }
```
The commit message has been changed accordingly.

QuerthDP commented on db/marshal/type_parser.cc:
```diff
@@ -14,11 +14,13 @@
 
 #include <stdexcept>
 #include <string>
+#include <tuple>
```
The commit message has been changed accordingly.

QuerthDP commented on test/boost/types_test.cc:
```diff
@@ -699,6 +699,17 @@ BOOST_AUTO_TEST_CASE(test_parse_invalid_tuple) {
     BOOST_REQUIRE_THROW(parser.parse(), exceptions::configuration_exception);
 }
 
+BOOST_AUTO_TEST_CASE(test_parse_valid_vector) {
+    auto parser = db::marshal::type_parser("org.apache.cassandra.db.marshal.VectorType(org.apache.cassandra.db.marshal.Int32Type,5)");
+    auto type = parser.parse();
+    BOOST_REQUIRE(type->as_cql3_type().to_string() == "vector<int, 5>");
+}
+
+BOOST_AUTO_TEST_CASE(test_parse_invalid_vector) {
+    auto parser = db::marshal::type_parser("org.apache.cassandra.db.marshal.VectorType()");
+    BOOST_REQUIRE_THROW(parser.parse(), exceptions::configuration_exception);
+}
+
```
The commit message has been changed accordingly.

QuerthDP commented on test/boost/cql_functions_test.cc:
```diff
@@ -23,6 +23,7 @@
 
 #include <seastar/core/future-util.hh>
 #include "transport/messages/result_message.hh"
+#include "types/vector.hh"
```
The commit message has been changed accordingly.

QuerthDP commented on transport/server.cc:
```diff
@@ -20,6 +20,7 @@
 #include "types/list.hh"
 #include "types/set.hh"
 #include "types/map.hh"
+#include "types/vector.hh"
```
The commit message has been changed accordingly.

Please check, if the message now is more explanatory.

QuerthDP commented on test/cqlpy/cassandra_tests/validation/entities/vectors_test.py:
```diff
@@ -0,0 +1,16 @@
+# This file was translated from the original Java test from the Apache
```
The commit message has been changed, previous version was added by mistake.

QuerthDP commented on cql3/expr/expression.cc:
```diff
@@ -24,6 +24,8 @@
 #include "types/list.hh"
 #include "types/map.hh"
 #include "types/set.hh"
+#include "types/types.hh"
```
It was not necessary indeed. Rebased.

QuerthDP commented on types/types.cc:
```diff
@@ -1609,6 +1631,143 @@ static void validate_aux(const tuple_type_impl& t, View v) {
+}
+
+bytes_view read_vector_element(bytes_view& v, std::optional<size_t> value_length_if_fixed) {
+    int32_t element_size = 0;
```
It was not needed to initialize the variable here. Works fine as expected without initialising it explicitly.
Changed via rebase.

QuerthDP commented on test/boost/expr_test.cc:
```diff
@@ -741,6 +753,41 @@ BOOST_AUTO_TEST_CASE(evaluate_tuple_constructor_with_prefix_fields) {
+BOOST_AUTO_TEST_CASE(evaluate_vector_constructor_empty) {
+    expression empty_vector = make_vector_constructor({}, int32_type, 3);
+    BOOST_REQUIRE_EQUAL(evaluate(empty_vector, evaluation_inputs{}), make_int_vector_raw({}));
+}
```
Removed.

QuerthDP commented on test/boost/expr_test.cc:
```diff
@@ -741,6 +753,41 @@ BOOST_AUTO_TEST_CASE(evaluate_tuple_constructor_with_prefix_fields) {
+        make_vector_constructor({make_int_const(1), make_empty_const(int32_type), make_int_const(3)}, int32_type, 3);
+    BOOST_REQUIRE_EQUAL(evaluate(vector_with_empty, evaluation_inputs{}),
+                        make_vector_raw({make_int_raw(1), make_empty_raw(), make_int_raw(3)}));
+}
```
Removed.

QuerthDP commented on concrete_types.hh:
```diff
@@ -15,6 +15,7 @@
 #include "types/map.hh"
 #include "types/set.hh"
 #include "types/tuple.hh"
+#include "types/vector.hh"
```
Changed the commit message accordingly.

Although I've left the from_string implementation here so far, because I'm not sure if it's needed at all and might be removed (see the discussion [below](https://github.com/zpp-2024-vector-search/scylladb/pull/5#discussion_r1898372571)).

janpiotrlakomy commented:
Changelog:
- removed from_string implementation for vectors
- added a vector size limit
- fixed a bug in type_parser::get_vector_parameters
- fixed vector serialization format for variable length elements to match Cassandra's implementation
- added a test for serialization of vectors with variable length elements.

Jadw1 commented:
Changelog:
I synced our master with Scylla master and rebased this PR on top of updated zpp-2024-vector-search/scylladb/master.
